packages/phone-number-privacy/common/src/utils/sequential-delay.ts

```diff
@@ -0,0 +1,80 @@
+import { SequentialDelayDomain, SequentialDelayStage } from '../interfaces'
+interface IndexedSequentialDelayStage extends SequentialDelayStage {
+  // The stage's index in the stage array
+  index: number
+  // The attempt number at which the stage begins
+  start: number
+}
+export interface SequentialDelayResult {
+  accepted: boolean
+  state?: SequentialDelayState
+}
+interface SequentialDelayState {
+  // Timestamp used for deciding when the next request will be accepted.
+  timer: number
+  // Number of queries that have been accepted for the SequentialDelayDomain instance.
+  counter: number
+}
+
+export const checkSequentialDelay = (
+  domain: SequentialDelayDomain,
+  attemptTime: number,
+  state?: SequentialDelayState
+): SequentialDelayResult => {
+  const counter = state?.counter ?? 0
+  const timer = state?.timer ?? 0
+  const stage = getIndexedStage(domain, counter)
+
+  if (!stage) {
+    return { accepted: false, state }
+  }
+
+  const resetTimer = stage.resetTimer ?? true
+  const delay = getDelay(stage, counter)
+  const notBefore = timer + delay
+
+  if (attemptTime < notBefore) {
+    return { accepted: false, state }
+  }
+
+  // Request is accepted. Update the state.
+  return {
+    accepted: true,
+    state: {
+      counter: counter + 1,
+      timer: resetTimer ? attemptTime : notBefore,
+    },
+  }
+}
+
+const getIndexedStage = (
+  domain: SequentialDelayDomain,
+  counter: number
+): IndexedSequentialDelayStage | null => {
+  let attemptsInStage = 0
+  let stage = 0
+  let i = 0
```

Comment:
Because stage and i must be iterated, I would have to create an iterable containing all their projected values and then loop through that in order to do this using only const (as far as I know). That doesn't sound like the cleaner approach to me personally. If there's another way to do this that I'm missing please let me know.

Reply:
I'm not seeing a way to do this other than creating an iterable of all the projected values for i and stage. Apologies if I'm missing something cleaner 😅